### PR TITLE
chore(warehouse): added merge partition pruning for deltalake changes.

### PR DESCRIPTION
**Pull request by achettyiitr**
# Description

Added support for Merge Partition pruning for MERGE Queries for Databricks. Here instead of just specifying the id column during the MERGE query, we will also specify the corresponding date range. This will boost up the Query and will only look for the specified partition.

Ref: https://kb.databricks.com/en_US/delta/delta-merge-into

## Notion Ticket

https://www.notion.so/rudderstacks/Deltalake-Partitioning-Pruning-36b3999668d244b39045c14d2a6aae5d
https://www.notion.so/rudderstacks/Add-support-for-dedup-handling-for-Partition-9df71fea56154b7c821bc5f353761571

Ref: https://kb.databricks.com/delta/delta-merge-into.html

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: warehouse/utils/utils_test.go [hunk 1]
```diff
@@ -1183,25 +1186,23 @@ func TestGetWarehouseIdentifier(t *testing.T) {
...
-			},
-		},
-	}
-	for _, input := range inputs {
-		got := JSONSchemaToMap(input.rawMsg)
-		require.Equal(t, got, input.expected)
-	}
-}
+var _ = Describe("Utils", func() {
+	DescribeTable("JSOM schema to Map", func(rawMsg json.RawMessage, expected map[string]map[string]string) {
```
Comment (abhimanyubabbar) [inline warehouse/utils/utils_test.go:1190]: JSON*

**Review by abhimanyubabbar — COMMENTED** [review]
Some cosmetic changes for the partition pruning in deltalake.

File: warehouse/utils/utils_test.go (hunk 1 above)
Comment (achettyiitr) [reply warehouse/utils/utils_test.go:1190]: Done.

File: warehouse/deltalake/query_builder.go [hunk 2]
```diff
@@ -33,12 +33,15 @@ func stagingSqlStatement(namespace, tableName, stagingTableName string, columnKe
 }
 
 // mergeableLTSQLStatement merge load table sql statement
-func mergeableLTSQLStatement(namespace, tableName, stagingTableName string, columnKeys []string) (sqlStatement string) {
+func mergeableLTSQLStatement(namespace, tableName, stagingTableName string, columnKeys []string, partitionedQuery string) (sqlStatement string) {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/deltalake/query_builder.go:36]: Done.

File: warehouse/deltalake/deltalake.go [hunk 3]
```diff
@@ -819,7 +872,7 @@ func (dl *HandleT) CreateSchema() (err error) {
 }
 
 // AlterColumn alter table with column name and type
-func (dl *HandleT) AlterColumn(tableName, columnName, columnType string) (err error) {
+func (dl *HandleT) AlterColumn(string, string, string) (err error) {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/deltalake/deltalake.go:875]: Done.

File: warehouse/deltalake/deltalake.go [hunk 4]
```diff
@@ -291,6 +294,54 @@ func (dl *HandleT) fetchTables(dbT *databricks.DBHandleT, sqlStatement string) (
...
+
+// partitionedQuery partition query
+func (dl *HandleT) partitionedQuery(tableName string) (partitionedQuery string) {
+	if !enablePartitionPruning {
+		return
+	}
+
+	partitionedColumns, err := dl.fetchPartitionColumns(dl.dbHandleT, tableName)
+	if err != nil {
+		pkgLogger.Errorf("error while fetching partitioned query with %s", err.Error())
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/deltalake/deltalake.go:328]: Done.

File: warehouse/deltalake/deltalake.go [hunk 5]
```diff
@@ -291,6 +294,54 @@ func (dl *HandleT) fetchTables(dbT *databricks.DBHandleT, sqlStatement string) (
...
+	return misc.ContainsString(partitionedColumns, "event_date")
+}
+
+// partitionedQuery partition query
+func (dl *HandleT) partitionedQuery(tableName string) (partitionedQuery string) {
+	if !enablePartitionPruning {
+		return
+	}
+
+	partitionedColumns, err := dl.fetchPartitionColumns(dl.dbHandleT, tableName)
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/deltalake/deltalake.go:326]: Done.

File: warehouse/deltalake/deltalake.go [hunk 6]
```diff
@@ -291,6 +294,54 @@ func (dl *HandleT) fetchTables(dbT *databricks.DBHandleT, sqlStatement string) (
...
+		Identifier:   dbT.CredIdentifier,
+		SqlStatement: sqlStatement,
+	})
+	if err != nil {
+		return partitionedColumns, fmt.Errorf("error while fetching partition columns with error: %w", err)
+	}
+	if !checkAndIgnoreAlreadyExistError(columnsResponse.GetErrorCode(), partitionNotFound) {
+		err = fmt.Errorf("error while fetching partition with response: %v", columnsResponse.GetErrorMessage())
+		return
+	}
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/deltalake/deltalake.go:311]: Done.

File: warehouse/deltalake/deltalake.go [hunk 7]
```diff
@@ -291,6 +294,54 @@ func (dl *HandleT) fetchTables(dbT *databricks.DBHandleT, sqlStatement string) (
...
+	partitionedColumns = append(partitionedColumns, columnsResponse.GetColumns()...)
+	return
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/deltalake/deltalake.go:313]: Done.

File: warehouse/deltalake/deltalake.go [hunk 8]
```diff
@@ -291,6 +294,54 @@ func (dl *HandleT) fetchTables(dbT *databricks.DBHandleT, sqlStatement string) (
...
+}
+
+func isPartitionedByEventDate(partitionedColumns []string) bool {
+	return misc.ContainsString(partitionedColumns, "event_date")
+}
+
+// partitionedQuery partition query
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/deltalake/deltalake.go:320]: Done.

File: warehouse/deltalake/deltalake.go [hunk 9]
```diff
@@ -291,6 +294,54 @@ func (dl *HandleT) fetchTables(dbT *databricks.DBHandleT, sqlStatement string) (
...
+func (dl *HandleT) partitionedQuery(tableName string) (partitionedQuery string) {
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/deltalake/deltalake.go:321]: Done.

File: warehouse/deltalake/deltalake.go [hunk 10]
```diff
@@ -291,6 +294,54 @@ func (dl *HandleT) fetchTables(dbT *databricks.DBHandleT, sqlStatement string) (
 	return
 }
 
+// fetchPartitionColumns fetch tables with tableNames
```
[the remark this answers is not in the document]
Comment (achettyiitr) [reply warehouse/deltalake/deltalake.go:297]: Done.

**Review by abhimanyubabbar — APPROVED** [review]
It looks good to me 👍🏼

**Review by Jayachand — APPROVED** [review]

**Pull request merged**